R/select_nse.R

```diff
@@ -0,0 +1,512 @@
+  first_obj <- .dynEval(expr[[2]], inherits = FALSE, minframe = 0L)
+  .eval_expr(
+    names(first_obj),
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = FALSE
+  )
+}
+
+# e.g starts_with("Sep")
+.select_helper <- function(expr, data, ignore_case, regex, verbose) {
+  lst_expr <- as.list(expr)
+
+  # need this if condition to distinguish between starts_with("Sep") (that we
+  # can use directly) and starts_with(i) (where we need to get i)
+  if (length(lst_expr) == 2L && typeof(lst_expr[[2]]) == "symbol") {
+    collapsed_patterns <- .dynGet(lst_expr[[2]], inherits = FALSE, minframe = 0L)
+  } else {
+    collapsed_patterns <- paste(unlist(lst_expr[2:length(lst_expr)]),
+      collapse = "|"
+    )
+  }
+
+  helper <- lst_expr[[1]]
+
+  rgx <- if (helper == "starts_with") {
+    paste0("^(", collapsed_patterns, ")")
+  } else if (helper == "ends_with") {
+    paste0("(", collapsed_patterns, ")$")
+  } else if (helper == "contains") {
+    paste0("(", collapsed_patterns, ")")
+  } else if (helper == "regex") {
+    collapsed_patterns
+  }
+  grep(rgx, colnames(data), ignore.case = ignore_case)
+}
+
+# e.g args$select (happens when we use grouped_data (see center.grouped_df()))
+.select_dollar <- function(expr, data, ignore_case, regex, verbose) {
+  first_obj <- .dynGet(expr[[2]], inherits = FALSE, minframe = 0L)
+  .eval_expr(
+    first_obj[[deparse(expr[[3]])]],
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+}
+
+# e.g ~ gear + cyl
+.select_tilde <- function(expr, data, ignore_case, regex, verbose) {
+  vars <- all.vars(expr)
+  unlist(lapply(
+    vars,
+    .eval_expr,
+    data = data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  ))
+}
+
+# e.g list(gear = 4, cyl = 5) [NOT SURE IT WILL BE USED]
+.select_list <- function(expr, data, ignore_case, regex, verbose) {
+  vars <- names(expr)
+  unlist(lapply(
+    vars,
+    .eval_expr,
+    data = data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  ))
+}
+
+# e.g is.numeric()
+.select_context <- function(expr, data, ignore_case, regex, verbose) {
+  x_dep <- insight::safe_deparse(expr)
+  if (endsWith(x_dep, "()")) {
+    new_expr <- gsub("\\(\\)$", "", x_dep)
+    new_expr <- str2lang(new_expr)
+    .eval_expr(
+      new_expr,
+      data = data,
+      ignore_case = ignore_case,
+      regex = regex,
+      verbose = verbose
+    )
+  } else {
+    .dynEval(expr, inherits = FALSE, minframe = 0L)
+  }
+}
+
+# -------------------------------------
+
+.check_data <- function(data) {
+  if (is.null(data)) {
+    insight::format_error("The `data` argument must be provided.")
+  }
+  .coerce_to_dataframe(data)
+}
+
+.regex_select_helper <- function() {
+  "(starts\\_with|ends\\_with|col\\_ends\\_with|contains|regex)"
+}
+
+.conflicting_packages <- function(packages = NULL) {
+  if (is.null(packages)) {
+    packages <- "poorman"
```

Comment:
Do we not need to include `{tidyselect}` here?

Reply:
I think all the `tidyverse` packages (and probably more) reexports helpers from `tidyselect` so we can't put all of them here. All tests run and they cover several use cases so I think it's fine